projects/pastanaga/src/lib/controls/checkbox-tree/checkbox-tree.component.ts

```diff
@@ -183,29 +187,48 @@ export class CheckboxTreeComponent implements ControlValueAccessor, OnInit {
         checkbox.isIndeterminate = this.getIndeterminateForFileSystem(checkbox);
         markForCheck(this.cdr);
         this.emitSelectionChanged();
+        this.childrenSelection.emit({parentId: checkbox.id, isSelected});
     }
 
-    private getUpdatedChildrenSelection(isSelected: boolean, checkbox: ControlModel): ControlModel[] {
-        return !checkbox.children ? [] : checkbox.children.map(child => {
-            if (!!child.children) {
-                child.children = this.getUpdatedChildrenSelection(isSelected, child);
-                child.selectedChildren = this.getSelectedChildrenCount(child);
-            }
-            return new ControlModel({...child, isSelected, isIndeterminate: false});
-        });
+    private getUpdatedChildrenSelection(isSelected: boolean, checkbox: ControlModel): ControlModel[] | undefined {
+        if (!!checkbox.children) {
+            return checkbox.children.map(child => {
+                if (!!child.children) {
+                    child.children = this.getUpdatedChildrenSelection(isSelected, child);
+                    child.selectedChildren = this.getSelectedChildrenCount(child);
+                }
+                return new ControlModel({...child, isSelected, isIndeterminate: false});
+            });
+        }
     }
 
     setParentState(childrenTree: ControlModel[], parent: ControlModel) {
-        parent.children = childrenTree;
+        // copy children state without changing object ref to prevent having an infinite loop of children/parent change detection
+        if (!!parent.children) {
+            parent.children.forEach((child, i) => {
+                const childModel = childrenTree[i];
+                if (!!childModel) {
+                    child.label = childModel.label;
```

Comment:
why not
```
child = {...childModel};
```
?

Reply:
to prevent having a new object triggering the changeDetection cycle (as I tried to explain in my comment but maybe I was unclear).
The thing is this function is called whenever the child tree is updated, so we can update the parent and bubble up the selection event. So the child is already in the right state, but having `parent.children = childrenTree ` was triggering the detection cycle anyway and it was starting a loop where children triggered parent who triggered child and so on.

ohhh! I see what you mean actually, changing the reference of each child in children without changing the reference of children should do the trick as well. I'll try tomorrow :)